"use client";
import { AxiosError } from "axios";

import axios from "axios";

const instance = axios.create({
  baseURL: "http://147.45.163.127:9000/api/",
});

instance.interceptors.request.use((config) => {
  const token = localStorage.getItem("access_token");
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

export const mainService = {
  setAuthHeader(token: string) {
    instance.defaults.headers.common["Authorization"] = `Bearer ${token}`;
  },

  async register(data: {
    password: string;
    email: string;
    phone: string;
    username: string;
  }) {
    try {
      const response = await instance.post("register", data);
      return response.data;
    } catch (error) {
      if (error instanceof AxiosError) {
        // üëà –ü—Ä–æ–≤–µ—Ä—è–µ–º, —è–≤–ª—è–µ—Ç—Å—è –ª–∏ –æ—à–∏–±–∫–∞ axios-–æ—à–∏–±–∫–æ–π
        throw new Error(
          error.response?.data?.detail?.ru || "–û—à–∏–±–∫–∞ —Ä–µ–≥–∏—Å—Ç—Ä–∞—Ü–∏–∏"
        );
      }
      throw new Error("–ù–µ–∏–∑–≤–µ—Å—Ç–Ω–∞—è –æ—à–∏–±–∫–∞");
    }
  },

  async login(data: { email: string; password: string }) {
    try {
      const response = await instance.post("/login", data);
      return response.data;
    } catch (error: any) {
      console.error("Error during login:", error);
      throw new Error(error.response?.data?.message || "–û—à–∏–±–∫–∞ –∞–≤—Ç–æ—Ä–∏–∑–∞—Ü–∏–∏");
    }
  },

  async verifyEmail(data: { email: string; code: string }) {
    if (!data.email) throw new Error("Email –Ω–µ –Ω–∞–π–¥–µ–Ω");

    try {
      const response = await instance.post("/verify", {
        email: data.email,
        code: data.code,
      });
      return response.data;
    } catch (error: any) {
      console.error("Error during email verification:", error);
      throw new Error(error.response?.data?.message || "–û—à–∏–±–∫–∞ –≤–µ—Ä–∏—Ñ–∏–∫–∞—Ü–∏–∏");
    }
  },

  async resetCode(data: { email: string }) {
    try {
      const response = await instance.post("/reset-code", data);
      return response.data;
    } catch (error: any) {
      throw new Error(error.response?.data?.message || "–û—à–∏–±–∫–∞ –∑–∞–ø—Ä–æ—Å–∞");
    }
  },

  async resetPassword(data: { email: string; code: string }) {
    try {
      const response = await instance.post("reset-pwd", {
        email: data.email,
        code: data.code,
      });
      return response.data;
    } catch (error) {
      console.error("Error during password reset:", error);
      throw error;
    }
  },
};
